### PR TITLE
Add iterations parameter to //smooth

### PR DESCRIPTION
**Pull request by sbrl**
Here's try #2!

This PR adds an iterations parameter to `//smooth`, as I've found myself wanting to run many passes at once.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -24,7 +24,7 @@ Meaning: Add grass on top and change deeper nodes to stone.
 Let the map generator generate ores in the selected region, this works with whatever ores mods have registered.
 The optional argument specifies which height is used for deciding the ores to generate and defaults to sea level (0).
 
-* `//smooth`
+* `//smooth [<iterations>]`
 
 Smooth terrain in the current selection. When adding nodes, the type of the top-most node in the column is used.
 This uses an exponentially weighted moving average (EWMA) and should be ran 1 or 2 times to produce good results.
```
Comment (sfan5) [inline README.md]: description should be updated
"should be ran with 1 or 2 iterations"

File: init.lua [hunk 2]
```diff
@@ -197,131 +197,133 @@ local function smooth(pos1, pos2, deadzone)
...
-	-- read heightmap from data
-	local heightmap = {}
-	local hstride = {x=1, z=dim.x}
-	for x = 0, dim.x-1 do
-		for z = 0, dim.z-1 do
-			heightmap[x + (z * hstride.z) + 1] = 0
+	
+	local count = 0
+	for i = 1,iterations do
+		-- read heightmap from data
```
Comment (sfan5) [inline init.lua]: reading/writing the map only needs to be done once

Comment (sbrl) [reply init.lua]: Ah, you're right. I didn't notice that! I'll move the heightmap calculation outside the loop.

**Comment by sbrl** [comment]
Thanks for the review! I've corrected those issues.

File: init.lua [hunk 3]
```diff
@@ -197,7 +197,7 @@ local function smooth(pos1, pos2, deadzone)
 	local offset = vector.subtract(pos1, area.MinEdge)
 	local c_air = minetest.get_content_id("air")
 	local c_dirt = minetest.get_content_id("default:dirt")
-
+
```
Comment (sfan5) [inline init.lua]: whitespace

File: init.lua [hunk 4]
```diff
@@ -221,107 +221,109 @@ local function smooth(pos1, pos2, deadzone)
...
+					local y = old_height
+					while y <= new_height-1 do
+						local index = index_z + (offset.y + y) * stride.y
+						if data[index] == c_air then data[index] = c_top end
 
-					count = count + 1
-					y = y + 1
+						count = count + 1
+						y = y + 1
+					end
```
Comment (sfan5) [inline init.lua]: writing it back to the map only needs to be done once too...

File: init.lua (hunk 3 above)
Comment (sbrl) [reply init.lua]: Erm you want it removed? Uhhh sure? Not my preference (I think my editor did it automagically), but I can change it here.

File: init.lua (hunk 4 above)
Comment (sbrl) [reply init.lua]: Ah, ok. I'll take a look.

**Comment by sbrl** [comment]
Fixed!

**Comment by sfan5** [comment]
I noticed that running just the algorithm multiple times doesn't really work out easily, so I've just implemented this myself.

**Comment by sbrl** [comment]
Ah, ok.....? What did you do differently? I'm curious as I've like to improve my own programming skills.

**Comment by sfan5** [comment]
In the current state of this PR, the code will read heights from the heightmap and compute slices in a loop.

There problem here is that later iterations would just read from the heightmap again, completely ignoring the previous results. The second problem is, the final heights are calculated inside the loop that "writes" heights back to the map, so that needs to be done multiple times too.

What I've done is to refactor the algorithm into a separate method that does both: compute slices and calculate the new heightmap.
Then I can easily call the method in a loop, each time reusing and amending the previous results.

**Comment by sbrl** [comment]
Ahhh, I see. Thanks for the explanation :smiley_cat:

**Pull request closed without merging**